fix: import accounts requiring 2 attempts for account to persist

woodenfurniture opened the pull request:
## Description

This PR includes the following changes:
1. Fixes issue where importing an account via account management requires 2 attempts in order for the new account to persist in redux.
2. A simplification of state management where we track `enabledAccountIds` rather than `hiddenAccountIds`, negating the need to check the existence of an account in state management before fetching it and toggling it (which created a race condition).
3. Improved handling of no accounts yet connected in the UI (see screenshots below)

Not included known issues to be followed up on:
* loading state during initial wallet connection briefly shows empty "add a chain to get started" 
* loading state of import accounts drawer borked, which is already raised in upstream PR and will be resolved
* loading state when clicking "done" does not exist yet

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #6818

## Risk
> High Risk PRs Require 2 approvals

High risk as it changes state management of portfolio and the logic of initial load of portfolio data.

> What protocols, transaction types or contract interactions might be affected by this PR?

## Testing

Check loading app from `previous session` and then `empty cache` both load portfolio as intended. Should be no functional change from prod.

Check importing an account doesn't require 2 attempts for your selection to persist in the account management drawer (see linked issue for video demonstrating the original issue)

### Engineering

<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

### Operations

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

## Screenshots (if applicable)

No accounts connected have improved UI:
![image](https://github.com/shapeshift/web/assets/125113430/cbb0580d-8d23-4efc-8155-ea6c33688600)

Accounts added show the same UI as previously:
![image](https://github.com/shapeshift/web/assets/125113430/f2c02c6f-8fdc-4ff7-a0c0-ec1a7dfde35e)

Demo of account import persisting with first attempt:

https://github.com/shapeshift/web/assets/125113430/62f779fe-38d9-4c7f-b04f-aa6629640451

woodenfurniture commented on src/context/AppProvider/AppContext.tsx:
```diff
@@ -154,6 +154,10 @@ export const AppProvider = ({ children }: { children: React.ReactNode }) => {
+
+      for (const accountId of Object.keys(accountMetadataByAccountId)) {
+        dispatch(portfolio.actions.enableAccountId(accountId))
+      }
```
high risk here

gomesalexandre replied:
Wrap this in a feature flag check @woodenfurniture

woodenfurniture commented on src/state/slices/portfolioSlice/portfolioSlice.ts:
```diff
@@ -137,17 +137,26 @@ export const portfolio = createSlice({
+     * Explicitly enable an account by its `AccountId`. Necessary where `use-strict` toggles twice
+     * during initial load, leading to all auto-detected accounts being disabled.
+     */
+    enableAccountId: (draftState, { payload: accountId }: { payload: AccountId }) => {
```
high risk here

0xApotheosis commented on src/state/slices/portfolioSlice/portfolioSlice.ts:
```diff
@@ -137,17 +137,26 @@ export const portfolio = createSlice({
-      const hiddenAccountIdsSet = new Set(draftState.hiddenAccountIds)
-      const isHidden = hiddenAccountIdsSet.has(accountId)
+    /**
+     * Explicitly enable an account by its `AccountId`. Necessary where `use-strict` toggles twice
+     * during initial load, leading to all auto-detected accounts being disabled.
```
gm idempotence

0xApotheosis reviewed (changes requested):
Possibly a regression when adding accounts the old fashioned way:

https://github.com/shapeshift/web/assets/97164662/ff99b721-b7bf-479c-a4c2-307c97939d55

Note, I was able to successfully add an account this way on the current `develop` environment, both with and without the account management feature flags enabled.

woodenfurniture commented:
> Possibly a regression when adding accounts the old fashioned way:
> 
>  cannot-add-account.mp4 
> Note, I was able to successfully add an account this way on the current `develop` environment, both with and without the account management feature flags enabled.

gm ser, well spotted. Fixed, rebased and pushed 🙏

Uploading Screen Recording 2024-05-07 at 5.09.57 PM.mov…

woodenfurniture commented on src/components/ManageAccountsDrawer/components/ImportAccounts.tsx:
```diff
@@ -161,12 +162,12 @@ export const ImportAccounts = ({ chainId, onClose }: ImportAccountsProps) => {
   const chainNamespaceDisplayName = asset?.networkName ?? ''
   const [autoFetching, setAutoFetching] = useState(true)
-  const [toggledActiveAccountIds, setToggledActiveAccountIds] = useState<Set<AccountId>>(new Set())
+  const [toggledAccountIds, setToggledAccountIds] = useState<Set<AccountId>>(new Set())
```
Ah, but this is a state variable tracking "toggled" account IDs, where they could be previously enabled, disabled, or new. The original implementation tracked enabled state but code review required I implemented it as toggles per #6816 . 

Feels weird yeah, but we have to track it this was to avoid having 2 paradigms of state tracking running at the same time.

0xApotheosis reviewed (approved):
That last commit did the trick, that regression is now fixed 🤝

gomesalexandre commented on src/components/Modals/ManageAccounts/ManageAccountsModal.tsx:
```diff
@@ -82,8 +82,7 @@ export const ManageAccountsModal = ({
 
-  // TODO: redux will have to be updated to use the selected chains from this flow
-  const walletSupportedChainIds = useAppSelector(selectWalletSupportedChainIds)
+  const walletConnectedChainIds = useAppSelector(selectWalletChainIds)
```
🧠

gomesalexandre commented on src/pages/Accounts/AddAccountModal.tsx:
```diff
@@ -89,9 +89,10 @@ export const AddAccountModal = () => {
-      )
+      accountIds.forEach(accountId => {
+        dispatch(getAccount.initiate({ accountId, upsertOnFetch: true }, opts))
+        dispatch(portfolio.actions.enableAccountId(accountId))
```
And check for flag enabled here too

gomesalexandre replied:
@woodenfurniture feels weird that https://github.com/shapeshift/web/pull/6819/commits/1b608a5ccb7d4b1190ff27863589a31b6c876078 fixed things here - shouldn't we gracefully handle things when the account management flags are off i.e no notion of enabled AccountIds?

woodenfurniture replied:
Likely not, unless we plan on removing the feature in the near future. Since the notion of enabled accountIds is part of our redux schema, toggling this behavior would likely increase the complexity and risk moreso than simply enabling them here and keeping things relatively simple.

gomesalexandre commented on src/state/slices/common-selectors.ts:
```diff
@@ -26,10 +26,10 @@ export const selectWalletSupportedChainIds = (state: ReduxState) =>
+  (walletId, walletById, enabledAccountIds): AccountId[] => {
     const walletAccountIds = (walletId && walletById[walletId]) ?? []
-    return walletAccountIds.filter(accountId => !hiddenAccountIds.includes(accountId))
+    return walletAccountIds.filter(accountId => enabledAccountIds.includes(accountId))
```
If we do add proper feature flag checks, this will cause a bunch of regressions across the app with the change from `!hiddenAccountIds.includes()` (non-breaking change as the array would be empty without hidden AccountIds) to `enabledAccountIds.includes()` (now assumes `enabledAccountIds`.

In the spirit of progressive, non-breaking enhancements here, we should probably check for feature flag enabled here as well and not run the predicate if the flag is off?

gomesalexandre replied:
Also see https://shapeshift-dao.sentry.io/issues/215187/?environment=localhost&query=is:unresolved&referrer=issue-stream&statsPeriod=1h&stream_index=0

woodenfurniture replied:
Yes, this is from the team toggling the feature on and off in production, not users. Apotheosis actaully created a PR to address this
https://github.com/shapeshift/web/pull/6840
But we closed it as it's only due to forward schema migration being forced backwards when toggling the feature off.

Ok, since this has come up twice I'll wrap everything in the feature flag in a follow up 👍

woodenfurniture replied:
Keep in mind the actions to enable accountIds will need to stay so toggling the feature on will show some accounts (not all hidden), so only checking the toggle here.

woodenfurniture replied:
On second thought, this will not make a difference once this PR goes into prod as the forward migration will have occurred for all users. I'll add a commit here to coalesce so a rollback will not rug all users in prod, and skip the feature flag here.

gomesalexandre reviewed (approved):
Tested locally, confirmed this does what this says on the box at runtime with no regressions:

- Confirmed the old Add Account modal is still happy with the flag off
- Confirmed the new account management modal now only requires one attempt to persist

https://github.com/shapeshift/web/assets/17035424/a61f35ff-fe08-404d-b904-92966e29248d

Added some comments re: change from `hiddenAccountIds` to `enabledAccountIds` which was added without a feature flag - while it *is* happy at runtime, feels like we should have proper feature flag checks in place until the feature is enabled.
